asf_welsh_energy_consultation/pipeline/process_data.py

```diff
@@ -0,0 +1,256 @@
+    """
+    mcs = get_mcs_domestic()
+    og = get_offgas()
+    countries = get_countries()
+    rural = get_rurality()
+
+    # join with off-gas data
+    mcs = mcs.merge(og, on="postcode", how="left")
+    mcs["off_gas"] = mcs["off_gas"].fillna("On gas").replace({True: "Off gas"})
+
+    # join with regions in order to filter to Wales
+    mcs = mcs.merge(countries, on="postcode", how="left")
+    mcs = mcs.loc[mcs["country"] == "Wales"].reset_index(drop=True)
+    # 1203 records with no match - 273 are Northern Ireland which leaves 918
+    # Some will be new postcodes (new build developments)
+    # and some may be expired postcodes
+    # In future, implement new solution that uses outward codes
+
+    # join with rurality data
+    mcs = mcs.merge(rural, on="postcode", how="left")
+    # only 13 postcodes lost in this merge
+
+    # add custom rurality column
+    mcs["rurality_7"] = mcs["rurality_10_label"].replace(
+        {
+            "Urban city and town": "Urban",
+            "Urban major conurbation": "Urban",
+            "Urban city and town in a sparse setting": "Urban",
+            "Urban minor conurbation": "Urban",
+        }
+    )
+
+    return mcs
+
+
+# load enhanced MCS as part of this script, so only needs to be done once
+mcs = get_enhanced_mcs()
+
+
+def cumsums_by_variable(variable, new_var_name, data=mcs):
+    """Process data into a form giving the cumulative total of
+    installations on each date for each category of a variable.
+
+    Args:
+        variable (str): Variable to split by.
+        new_var_name (str): Name of variable in processed dataset.
+        data (pd.DataFrame, optional): Base data. Defaults to mcs.
+
+    Returns:
+        pd.DataFrame: Cumulative totals dataset.
+    """
+
+    # calculate total number of installations for each date/category pair
+    totals = data.groupby(["commission_date", variable]).size()
+
+    totals = totals.reset_index().rename(columns={0: "sum"})
+
+    idx = pd.date_range(
+        totals["commission_date"].min(), totals["commission_date"].max()
+    )
+
+    totals = totals.pivot(index="commission_date", columns=variable).fillna(0)
+
+    totals.index = pd.DatetimeIndex(totals.index)
+
+    # reindex, fill in gaps with 0, and take cumulative sums
+    cumsums = totals.reindex(idx, fill_value=0).cumsum()
+    cumsums.columns = cumsums.columns.droplevel(0)
+    cumsums = cumsums.rename_axis(None, axis=1).reset_index(names="date")
+
+    cumsums = pd.melt(
+        cumsums,
+        id_vars="date",
+        value_vars=[col for col in cumsums.columns if col != "date"],
+    ).rename(columns={"variable": new_var_name, "value": "Number of heat pumps"})
+
+    cumsums["Number of heat pumps"] = cumsums["Number of heat pumps"].astype(int)
+
+    # remove pre-2015 entries as numbers are small
+    cumsums = cumsums.loc[cumsums.date >= "2015-01-01"].reset_index(drop=True)
+
+    return cumsums
+
+
+# PROCESSING EPC
+
+wales_epc = get_wales_epc()
```

Comment:
Because this takes a long time and `get_wales_epc_new` is called twice, I wondered if it might be quicker to put it outside the function so it gets loaded when the script gets imported. If that's how it works...

Reply:
I think that's fine, yes :) keep it as is!